thentos-core/src/Network/HostAddr.hs

```diff
@@ -0,0 +1,33 @@
+{-# LANGUAGE LambdaCase #-}
+
+module Network.HostAddr
+where
+
+import Data.Int (Int32)
+import Network.Socket
+          (SockAddr(SockAddrInet, SockAddrInet6, SockAddrUnix, SockAddrCan),
+           HostAddress, HostAddress6, SocketType(Stream), AddrInfoFlag(AI_NUMERICHOST),
+           addrAddress, getAddrInfo, addrFlags, addrSocketType, defaultHints)
+
+data HostAddr
+    = HostAddress  HostAddress
+    | HostAddress6 HostAddress6
+    | UnixAddress  String
+    | CanAddress   Int32
+    deriving (Eq, Read, Show)
+
+hostAddr :: SockAddr -> HostAddr
+hostAddr = \case
+    SockAddrInet _ ip -> HostAddress ip
+    SockAddrInet6 _ _ ip _ -> HostAddress6 ip
+    SockAddrUnix fp -> UnixAddress fp
+    SockAddrCan i -> CanAddress i
+
+getHostAddr :: String -> IO HostAddr
+getHostAddr name = hostAddr . addrAddress . single <$> getAddrInfo (Just hints) (Just name) Nothing
+  where
+    hints = defaultHints { addrFlags = [AI_NUMERICHOST], addrSocketType = Stream }
+
+    single [x] = x
+    single []  = error "Impossible error: numeric addresses should always resolve"
```

Comment:
[Your last rebase killed my earlier comments, so I'll add it to the "changed files" tab and not to the commit this time.]

"impossible error" means "programmer failed", but this is an instance of "configuration file is bad" (at least potentially).  I would prefer:

```haskell
single [] = throwIO $ ErrorCall "Configuration error: numeric address should always resolve"
```

or even change the result type to `Either String HostAddr` and deal with the error at the call site.

I prefer `throwIO` to `error` even if the exception type is no more specific, because it works without the implicit `unsafePerformIO`.  Does that make sense?